### PR TITLE
One sided constraint support for HPIPM solvers

### PR DESCRIPTION
**Pull request by FreyJo**
- add `ACADOS_INFTY` in Python and C, `get_acados_infty` in Matlab.
- If any constraints have bounds `ACADOS_INFTY`, the constraint is marked as not relevant in the QP representation object `ocp_qp`, by setting the corresponding `d_mask` values to `0`.
- The underlying QP solver interface should make use of this mask.
- The HPIPM solvers do make use of this natively.
- For now the value is `ACADOS_INFTY=1e10` which might can be changed later after some more extensive testing.
- Support for that in QP solvers other than HPIPM is not included in this PR and tracked in https://github.com/acados/acados/issues/650

### CODE REVIEW

File: acados/ocp_nlp/ocp_nlp_constraints_bgh.c [hunk 1]
```diff
@@ -1644,6 +1650,24 @@ void ocp_nlp_constraints_bgh_update_qp_vectors(void *config_, void *dims_, void
     // fun[2*ni : 2*(ni+ns)] = - slack + slack_bounds
     blasfeo_daxpy(2*ns, -1.0, memory->ux, nu+nx, &model->d, 2*nb+2*ng+2*nh, &memory->fun, 2*nb+2*ng+2*nh);
 
+    // Set dmask for QP: 0 means unconstrained.
+    for (int i = 0; i < nb+ng+nh; i++)
+    {
+        if (BLASFEO_DVECEL(&model->d, i) == -ACADOS_INFTY)
```
Comment (sandmaennchen) [inline acados/ocp_nlp/ocp_nlp_constraints_bgh.c:1656]: `<=`

File: acados/ocp_nlp/ocp_nlp_constraints_bgh.c [hunk 2]
```diff
@@ -1644,6 +1650,24 @@ void ocp_nlp_constraints_bgh_update_qp_vectors(void *config_, void *dims_, void
...
+        {
+            // printf("found upper infinity bound\n");
+            BLASFEO_DVECEL(memory->dmask, i) = 0;
+        }
+    }
+    for (int i = nb+ng+nh; i < 2*nb+ng+nh; i++)
+    {
+        if (BLASFEO_DVECEL(&model->d, i) == ACADOS_INFTY)
```
Comment (sandmaennchen) [inline acados/ocp_nlp/ocp_nlp_constraints_bgh.c:1664]: `>=`

**Review by sandmaennchen — APPROVED** [review]

File: acados/ocp_nlp/ocp_nlp_constraints_bgh.c [hunk 3]
```diff
@@ -1644,6 +1653,27 @@ void ocp_nlp_constraints_bgh_update_qp_vectors(void *config_, void *dims_, void
     // fun[2*ni : 2*(ni+ns)] = - slack + slack_bounds
     blasfeo_daxpy(2*ns, -1.0, memory->ux, nu+nx, &model->d, 2*nb+2*ng+2*nh, &memory->fun, 2*nb+2*ng+2*nh);
 
+    // Set dmask for QP: 0 means unconstrained.
+    for (int i = 0; i < nb+ng+nh; i++)
+    {
+        if (BLASFEO_DVECEL(&model->d, i) <= -ACADOS_INFTY)
```
Comment (giaf) [inline acados/ocp_nlp/ocp_nlp_constraints_bgh.c:1659]: I would go for exactly == , never know if a user wants to solve a badly scaled problem on purpose.
At least exactly == decreases likelihood of removing some desired constraint by chance.

Comment (giaf) [reply acados/ocp_nlp/ocp_nlp_constraints_bgh.c:1659]: Unless there are reasons why it should not hold exactly == with a constraint that was set to ACADOS_INFTY.

**Comment by giaf** [comment]
`d_mask` has the same size as `d` (namely `2*nb+2*ng+2*ns`) in HPIPM, and also `model->d` has the same (linearized) size in constraints bgh and bgp.
So also the term `+2*ns` should be considered.

**Comment by giaf** [comment]
HPIPM is implemented such that soft constraints slacks in the primal variables, Lagange multipliers and relative slacks are guaranteed to be exactly 0 for disregarded constraints, see e.g. https://github.com/giaf/hpipm/blob/master/ocp_qp/x_ocp_qp_ipm.c#L1821.
This means that at the NLP level you just need to mask out final components in the `fun` evaluation, as you are doing.

However, in the case of another QP solver, it has to be guaranteed in the QP solver interface that such quantities of disregarded constraints are and remain exactly at 0.
Otherwise, also input arguments to e.g. Hessian evaluation or fun evaluation or the like should be masked out properly.

**Comment by FreyJo** [comment]
Thanks for checking @giaf !
For now, this is only tested for HPIPM as mentioned in the PR descriptions.
In case of other QP solvers, we surely need to do more work and testing, also with soft constrained problems.
I would try to keep track of the current state on this and the potential issues you mentioned in https://github.com/acados/acados/issues/650

Could you elaborate on what you mean with "relative slacks" here:

> HPIPM is implemented such that [...] relative slacks are guaranteed to be exactly 0 for disregarded constraints  [...] 

Regarding the comparison with `ACADOS_INFTY`, I discussed more with @sandmaennchen and we decided to stick with the `>=` comparison.
If a user wants to solve a badly scaled problem on purpose, they still need to be aware that `ACADOS_INFTY` exists and that it is set to `1e10`.
They might have a constraint function bounded by `1e10` which would be ignored, and one with `5e10` which is included in the acados NLP, which is kind of weird behavior.
I also believe that this is not really a breaking change for anyone, as I doubt there are successful acados applications around with constraint bounds `> 1e10`.

**Pull request merged**